File: tutorials/testing/c++/circ_dllist.cpp
/**
 * Created by Kalpathi Subramanian, 2/20/18
 * krs@uncc.edu
 */

#include "Bridges.h"
#include "CircDLelement.h"
#include <string>


using namespace bridges;
using std::string;


int main(int argc, char **argv) {
	// create Bridges object
	Bridges bridges (YOUR_ASSSIGNMENT_NUMBER, "YOUR_USER_ID",
		"YOUR_API_KEY");

	CircDLelement<string>  *el0 = new CircDLelement<string> ("Gretel Chaney", "Gretel Chaney");
	CircDLelement<string>  *el1 = new CircDLelement<string> ("Lamont Kyler", "Lamont Kyler");
	CircDLelement<string>  *el2 = new CircDLelement<string> ("Gladys Serino", "Gladys Serino");
	CircDLelement<string>  *el3 = new CircDLelement<string> ("Karol Soderman", "Karol Soderman");
	CircDLelement<string>  *el4 = new CircDLelement<string> ("Starr McGinn", "Starr McGinn");

	// create the list
	el0->setNext(el1);
	el1->setPrev(el0);
	el1->setNext(el2);
	el2->setPrev(el1);
	el2->setNext(el3);
	el3->setPrev(el2);
	el3->setNext(el4);
	el4->setPrev(el3);

	// add the links to join the last and first nodes
	// must set the next and prev links
	el4->setNext(el0);
	el0->setPrev(el4);

	cout << "Testing forward list iterator.." << endl;
	CircDLelement_List<string> list(el0);
    auto iter = list.begin();
    do {
        cout << "\t" << *iter << endl;
        ++iter;
    } while (iter != list.end());

	cout << "Testing reverse list iterator.." << endl;
    iter = list.rbegin();
    do {
        cout << "\t" << *iter << endl;
        --iter;
    } while (iter != list.rend());

	// add  element colors
	// set colors for list elements - see the Color class for supported colors
	el0->setColor("red");
	el2->setColor("aliceblue");

	// color the links - must specify a valid terminating element
	el0->getLinkVisualizer(el1)->setColor("green");
	el3->getLinkVisualizer(el4)->setColor("magenta");

	// adjust link thickness
	el0->getLinkVisualizer(el1)->setThickness(2.0f);

	// set link label
	el2->getLinkVisualizer(el3)->setLabel("Link Label");

	// set node transparency
	el4->setOpacity (0.5f);

	// set node size
	el0->setSize (20);

	bridges.setDataStructure(el0);
	bridges.visualize();

	// freeing memory
	delete el0;
	delete el1;
	delete el2;
	delete el3;
	delete el4;
				

	return 0;
}

File: tutorials/testing/python/circ_dllist.py
from bridges.bridges import *
from bridges.circ_dl_element import *
import sys

def main():
    # create the Bridges object, set credentials
    bridges = Bridges(YOUR_ASSSIGNMENT_NUMBER, "YOUR_USER_ID", "YOUR_API_KEY")

    # title, description
    bridges.set_title("A Circular Doubly Linked List Example")
    bridges.set_description("This example shows circular doubly linked example with five nodes with assigned visual attributes")

	# create list elements
    el0 = CircDLelement(e="Gretel Chaney", label="Gretel Chaney")
    el1 = CircDLelement(e="Lamont Kyler", label="Lamont Kyler")
    el2 = CircDLelement(e="Gladys Serino", label="Gladys Serino")
    el3 = CircDLelement(e="Karol Soderman", label="Karol Soderman")
    el4 = CircDLelement(e="Starr McGinn", label="Starr McGinn")

    # link the elements
    el0.next = el1
    el1.prev = el0
    el1.next = el2
    el2.prev = el1
    el2.next = el3
    el3.prev = el2
    el3.next = el4
    el4.prev = el3

	# join first and last nodes
    el4.next = el0
    el0.prev = el4

	# visual attributes
    el0.visualizer.color = "red"
    el2.visualizer.color = 'aliceblue'

    el0.get_link_visualizer(el1).color = "green"
    el3.get_link_visualizer(el4).color = "magenta"

    el0.get_link_visualizer(el1).thickness = 2.0

    el4.opacity = 0.5

    el0.size = 20

    bridges.set_data_structure(el1)
    bridges.visualize()


if __name__ == "__main__":
    main()

